src/baal/active/heuristics/heuristics.py

```diff
@@ -241,27 +272,70 @@ class BatchBALD(BALD):
         Not tested on 4+ dims.
         """
 
-    def __init__(self, num_samples, shuffle_prop=0.0, reduction='none'):
+    def __init__(self, num_samples, num_draw=500, shuffle_prop=0.0, reduction='none'):
         self.epsilon = 1e-5
         self.num_samples = num_samples
+        self.num_draw = num_draw
         super().__init__(shuffle_prop=shuffle_prop,
                          reduction=reduction)
 
+    def _draw_choices(self, probs, n_choices):
+        """
+        Draw `n_choices` sample from `probs`.
+
+        Returns:
+            choices: B... x `n_choices`
+        """
+        probs = probs.permute(0, 2, 1)
+        probs_B_C = probs.reshape((-1, probs.shape[-1]))
+
+        # samples: Ni... x draw_per_xx
+        choices = torch.multinomial(probs_B_C,
+                                    num_samples=n_choices, replacement=True)
+
+        choices_b_M = choices.reshape(list(probs.shape[:-1]) + [n_choices])
+        return choices_b_M.long()
+
+    def _sample_from_history(self, probs, num_draw=1000):
```

Comment:
And a reference to https://github.com/BlackHC/BatchBALD/blob/master/src/joint_entropy/sampling.py ?